From 05c668aa82789f46b2f48ac642eb822b546faf22 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Mon, 20 Jul 2026 17:46:22 +0200
Subject: [PATCH 68/90] usb: dwc3: Track the power state of usb3_generic_phy

The USB3 PHYs will need to be powered off and on again from within the
set_mode work, independently of the rest of the controller. There is no
way to recover from a failure there, so track whether the PHYs are
actually powered to keep the phy_power_on()/phy_power_off() calls
balanced.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/usb/dwc3/core.c | 10 ++++++++--
 drivers/usb/dwc3/core.h |  6 ++++++
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index ad52e203563b..3d669323b40d 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -915,6 +915,8 @@ static int dwc3_phy_power_on(struct dwc3 *dwc)
 			goto err_power_off_usb3_phy;
 	}
 
+	dwc->usb3_phy_powered = true;
+
 	return 0;
 
 err_power_off_usb3_phy:
@@ -935,8 +937,12 @@ static void dwc3_phy_power_off(struct dwc3 *dwc)
 {
 	int i;
 
-	for (i = 0; i < dwc->num_usb3_ports; i++)
-		phy_power_off(dwc->usb3_generic_phy[i]);
+	if (dwc->usb3_phy_powered) {
+		for (i = 0; i < dwc->num_usb3_ports; i++)
+			phy_power_off(dwc->usb3_generic_phy[i]);
+
+		dwc->usb3_phy_powered = false;
+	}
 
 	for (i = 0; i < dwc->num_usb2_ports; i++)
 		phy_power_off(dwc->usb2_generic_phy[i]);
diff --git a/drivers/usb/dwc3/core.h b/drivers/usb/dwc3/core.h
index e0dee9d28740..19a93c3dd668 100644
--- a/drivers/usb/dwc3/core.h
+++ b/drivers/usb/dwc3/core.h
@@ -1067,6 +1067,10 @@ struct dwc3_glue_ops {
  * @usb3_generic_phy: pointer to array of USB3 PHYs
  * @num_usb2_ports: number of USB2 ports
  * @num_usb3_ports: number of USB3 ports
+ * @usb3_phy_powered: flag to indicate that the USB3 PHYs are powered on. The
+ *		      USB3 PHYs may be powered off independently of the rest of
+ *		      the controller, so their state has to be tracked to keep
+ *		      the phy_power_on()/phy_power_off() calls balanced.
  * @phys_ready: flag to indicate that PHYs are ready
  * @ulpi: pointer to ulpi interface
  * @ulpi_ready: flag to indicate that ULPI is initialized
@@ -1233,6 +1237,8 @@ struct dwc3 {
 	u8			num_usb2_ports;
 	u8			num_usb3_ports;
 
+	bool			usb3_phy_powered;
+
 	bool			phys_ready;
 
 	struct ulpi		*ulpi;
-- 
2.55.0

